File: DSA/Bhavya_vyas500121193/Day27_q1.py
class Node:
    def __init__(self, data): 
        self.data = data
        self.prev = None
        self.next = None

def delete_node(head, x):
    if head is None or x <= 0:
        return head

    current = head

    for _ in range(x - 1):
        if current is None:
            return head  
        current = current.next

    if current is None:
        return head

    if current.prev is None:
        head = current.next
        if head:
            head.prev = None
    else:
        current.prev.next = current.next

    if current.next:
        current.next.prev = current.prev

    return head
def print_list(head):
    while head:
        print(head.data, end=" <--> " if head.next else "\n")
        head = head.next

def build_doubly_linked_list(values):
    head = None
    prev = None
    for val in values:
        node = Node(val)
        if not head:
            head = node
        else:
            prev.next = node
            node.prev = prev
        prev = node
    return head


head1 = build_doubly_linked_list([1, 3, 4])
print("Before:")
print_list(head1)
head1 = delete_node(head1, 3)
print("After:")
print_list(head1)


head2 = build_doubly_linked_list([1, 5, 2, 9])
print("Before:")
print_list(head2)
head2 = delete_node(head2, 1)
print("After:")
print_list(head2)

File: DSA/Shlok_590017851/Day27_q1.c
#include <stdio.h>
#include <stdlib.h>

struct Node 
{
    int data;
    struct Node* prev;
    struct Node* next;
};
struct Node* createNode(int data)
{
    struct Node* newNode = (struct Node*) malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = NULL;
    return newNode;
}
struct Node* deleteNode(struct Node* head, int x) 
{
    if (head == NULL || x <= 0)
        return head;

    struct Node* temp = head;

    if (x == 1)
    {
        head = head->next;
        if (head != NULL)
            head->prev = NULL;
        free(temp);
        return head;
    }
    for (int i = 1; temp != NULL && i < x; i++) 
    {
        temp = temp->next;
    }
    if (temp == NULL)
        return head;

    if (temp->prev != NULL)
        temp->prev->next = temp->next;

    if (temp->next != NULL)
        temp->next->prev = temp->prev;

    free(temp);
    return head;
}
void printList(struct Node* head) 
{
    struct Node* temp = head;
    while (temp != NULL)
      {
        printf("%d", temp->data);
        if (temp->next != NULL)
            printf(" <--> ");
        temp = temp->next;
    }
    printf("\n");
}
int main()
{
    struct Node* head = createNode(1);
    struct Node* second = createNode(5);
    struct Node* third = createNode(2);
    struct Node* fourth = createNode(9);

    head->next = second;
    second->prev = head;
    second->next = third;
    third->prev = second;
    third->next = fourth;
    fourth->prev = third;
    printf("Original List:\n");
    printList(head);

    int x = 1; 
    head = deleteNode(head, x);
    printf("After deleting node at position %d:\n", x);
    printList(head);
    return 0;
}
